akka-http-core/src/main/scala-2.12/akka/http/scaladsl/util/FastFuture.scala

```diff
@@ -0,0 +1,122 @@
+  val failed: Throwable ⇒ Future[Nothing] = ErrorFuture.apply
+
+  private case class FulfilledFuture[+A](a: A) extends Future[A] {
+    def value = Some(Success(a))
+    def onComplete[U](f: Try[A] ⇒ U)(implicit executor: ExecutionContext) = Future.successful(a).onComplete(f)
+    def isCompleted = true
+    def result(atMost: Duration)(implicit permit: CanAwait) = a
+    def ready(atMost: Duration)(implicit permit: CanAwait) = this
+    def transform[S](f: scala.util.Try[A] ⇒ scala.util.Try[S])(implicit executor: scala.concurrent.ExecutionContext): scala.concurrent.Future[S] =
+      FastFuture(f(Success(a)))
+    def transformWith[S](f: scala.util.Try[A] ⇒ scala.concurrent.Future[S])(implicit executor: scala.concurrent.ExecutionContext): scala.concurrent.Future[S] =
+      new FastFuture(this).transformWith(f)
+  }
+  private case class ErrorFuture(error: Throwable) extends Future[Nothing] {
+    def value = Some(Failure(error))
+    def onComplete[U](f: Try[Nothing] ⇒ U)(implicit executor: ExecutionContext) = Future.failed(error).onComplete(f)
+    def isCompleted = true
+    def result(atMost: Duration)(implicit permit: CanAwait) = throw error
+    def ready(atMost: Duration)(implicit permit: CanAwait) = this
+    def transform[S](f: scala.util.Try[Nothing] ⇒ scala.util.Try[S])(implicit executor: scala.concurrent.ExecutionContext): scala.concurrent.Future[S] =
+      FastFuture(f(Failure(error)))
+    def transformWith[S](f: scala.util.Try[Nothing] ⇒ scala.concurrent.Future[S])(implicit executor: scala.concurrent.ExecutionContext): scala.concurrent.Future[S] =
+      new FastFuture(this).transformWith(f)
+  }
+
+  implicit class EnhancedFuture[T](val future: Future[T]) extends AnyVal {
+    def fast: FastFuture[T] = new FastFuture[T](future)
+  }
+
+  def sequence[T, M[_] <: TraversableOnce[_]](in: M[Future[T]])(implicit cbf: CanBuildFrom[M[Future[T]], T, M[T]], executor: ExecutionContext): Future[M[T]] =
+    in.foldLeft(successful(cbf(in))) {
+      (fr, fa) ⇒ for (r ← fr.fast; a ← fa.asInstanceOf[Future[T]].fast) yield r += a
+    }.fast.map(_.result())
+
+  def fold[T, R](futures: TraversableOnce[Future[T]])(zero: R)(f: (R, T) ⇒ R)(implicit executor: ExecutionContext): Future[R] =
+    if (futures.isEmpty) successful(zero)
+    else sequence(futures).fast.map(_.foldLeft(zero)(f))
+
+  def reduce[T, R >: T](futures: TraversableOnce[Future[T]])(op: (R, T) ⇒ R)(implicit executor: ExecutionContext): Future[R] =
+    if (futures.isEmpty) failed(new NoSuchElementException("reduce attempted on empty collection"))
+    else sequence(futures).fast.map(_ reduceLeft op)
+
+  def traverse[A, B, M[_] <: TraversableOnce[_]](in: M[A])(fn: A ⇒ Future[B])(implicit cbf: CanBuildFrom[M[A], B, M[B]], executor: ExecutionContext): Future[M[B]] =
+    in.foldLeft(successful(cbf(in))) { (fr, a) ⇒
+      val fb = fn(a.asInstanceOf[A])
+      for (r ← fr.fast; b ← fb.fast) yield r += b
+    }.fast.map(_.result())
+}
```

Comment:
IntelliJ claims this is identical to the other FastFuture? Oh wait, there is a third one...